src/Chainweb/Version.hs

```diff
@@ -209,109 +206,76 @@ instance IsMerkleLogEntry ChainwebHashTag ChainwebVersion where
-
--- test versions
-chainwebVersionToText v@Test{} = "test-" <> sshow (chainwebVersionId v)
-chainwebVersionToText v@TestWithTime{} = "testWithTime-" <> sshow (chainwebVersionId v)
-chainwebVersionToText v@TestWithPow{} = "testWithPow-" <> sshow (chainwebVersionId v)
-chainwebVersionToText v@Simulation{} = "simulation-" <> sshow (chainwebVersionId v)
+chainwebVersionToText v =
+    case HM.lookup v prettyVersions of
+        Just t -> t
+        Nothing -> case v of
+            Test{} -> "test"
+            TestWithTime{} -> "testWithTime"
+            TestWithPow{} -> "testWithPow"
+            Testnet00 -> "testnet00"
 {-# INLINABLE chainwebVersionToText #-}
 
--- | Read textual representation of Chainweb Version
+-- | Read textual representation of a `ChainwebVersion`.
 --
 chainwebVersionFromText :: MonadThrow m => T.Text -> m ChainwebVersion
-
--- Production versions
---
-chainwebVersionFromText "testnet00" = return Testnet00
-
--- Well-known test version names.
---
--- These are only used for parsing textual representations. There is a very low
--- chance that a roundtrip test for the 'HasTextRepresentation' of
--- 'ChainwebVersion' will succeed due to these names.
---
-chainwebVersionFromText "test" = return $ Test petersonChainGraph
-chainwebVersionFromText "test-singleton" = return $ Test singletonChainGraph
-chainwebVersionFromText "test-peterson" = return $ Test petersonChainGraph
-
-chainwebVersionFromText "testWithTime" = return $ TestWithTime petersonChainGraph
-chainwebVersionFromText "testWithTime-singleton" = return $ TestWithTime singletonChainGraph
-chainwebVersionFromText "testWithTime-peterson" = return $ TestWithTime petersonChainGraph
-
-chainwebVersionFromText "testWithPow" = return $ TestWithPow petersonChainGraph
-chainwebVersionFromText "testWithPow-singleton" = return $ TestWithPow singletonChainGraph
-chainwebVersionFromText "testWithPow-peterson" = return $ TestWithPow petersonChainGraph
-
--- Generic test versions
---
-chainwebVersionFromText t = case T.breakOnEnd "-" t of
-    (_, i) -> case treadM i of
-        Left e -> throwM
-            $ TextFormatException $ "Unknown Chainweb version: \"" <> t <> "\": " <> sshow e
-        Right x -> return $ fromTestChainwebVersionId x
-{-# INLINABLE chainwebVersionFromText #-}
+chainwebVersionFromText t =
+    case HM.lookup t chainwebVersions of
+        Just v -> pure v
+        Nothing -> case t of
+            "test" -> pure $ Test petersonChainGraph
+            "testWithTime" -> pure $ TestWithTime petersonChainGraph
+            "testWithPow" -> pure $ TestWithPow petersonChainGraph
+            _ -> throwM . TextFormatException $ "Unknown Chainweb version: " <> t
 
 instance HasTextRepresentation ChainwebVersion where
     toText = chainwebVersionToText
     {-# INLINE toText #-}
     fromText = chainwebVersionFromText
     {-# INLINE fromText #-}
 
+chainwebVersions :: HM.HashMap T.Text ChainwebVersion
+chainwebVersions = HM.fromList $
```

Comment:
This builds a map for every possible combination of version stem and graph. Would it be possible to reduce the map to just the choices of graphs and build the combined test chainweb version on the fly?

Reply:
hmm, maybe it doesn't matter that much... it's only a factor of three...